src/FsAutoComplete.Core/State.fs

```diff
@@ -38,6 +41,16 @@ type State =
     x.Files.[file] <- fileState
     x.FileCheckOptions.[file] <- opts
 
+  member x.AddCancellationToken(file : SourceFilePath, token: CancellationTokenSource) =
+    let lst = x.CancellationTokens.GetOrAdd(file, fun _ -> [])
+    let lst' = token::lst
+    x.CancellationTokens.[file] <- lst'
```

Comment:
`AddOrUpdate` to prevent a race?

(ideally we would have an atomic "cancel and clear" operation as well, see my comment above)